consensus/istanbul/backend/announce.go

```diff
@@ -288,54 +223,52 @@ func (sb *Backend) handleIstAnnounce(payload []byte) error {
 	}
 
 	if !regAndActiveVals[msg.Address] {
-		sb.logger.Warn("Received an IstanbulAnnounce message from a non registered validator. Ignoring it.", "AnnounceMsg", msg.String(), "validators", regAndActiveVals, "err", err)
+		sb.logger.Warn("Received an IstanbulAnnounce message from a non registered validator. Ignoring it.", "IstanbulMsg", msg.String(), "validators", regAndActiveVals, "err", err)
 		return errUnauthorizedAnnounceMessage
 	}
 
-	// Decrypt the EnodeURL
-	nodeKey := ecies.ImportECDSA(sb.GetNodeKey())
-
-	encryptedEndpoint := []byte("")
-	for _, entry := range msg.EncryptedEndpointData {
-		if bytes.Equal(entry[0], sb.Address().Bytes()) {
-			encryptedEndpoint = entry[1]
-		}
-	}
-	endpointBytes, err := nodeKey.Decrypt(encryptedEndpoint, nil, nil)
-	if err != nil && len(encryptedEndpoint) > 0 {
-		sb.logger.Warn("Error in decrypting endpoint", "err", err, "encryptedEndpoint", encryptedEndpoint)
+	var announceMessage announceMessage
+	err = rlp.DecodeBytes(msg.Msg, &announceMessage)
+	if err != nil {
+		sb.logger.Error("Error in decoding received Istanbul Announce message content", "err", err, "IstanbulMsg", msg.String())
+		return err
 	}
-	enodeUrl := msg.IncompleteEnodeURL + string(endpointBytes)
 
 	// Save in the valEnodeTable if mining
 	if sb.coreStarted {
-		block := sb.currentBlock()
-		valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
-
-		newValEnode := &validatorEnode{enodeURL: enodeUrl, view: msg.View}
-		if err := sb.valEnodeTable.upsert(msg.Address, newValEnode, valSet, sb.Address()); err != nil {
-			sb.logger.Warn("Error in upserting a valenode entry", "AnnounceMsg", msg, "error", err)
-			return err
+		var enodeUrl string
+		for _, entry := range announceMessage.EncryptedEnodeURLs {
+			if bytes.Equal(entry[0], sb.Address().Bytes()) {
```

Comment:
If it's not too tricky it would be more readable for the entries to be a struct, e.g. `entry.Address`, `entry.EncryptedEnodeUrl`

Reply:
I created a `AnnounceRecord` type.